netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java

```diff
@@ -239,12 +275,18 @@ NettyChannelBuilder eventLoopGroupPool(ObjectPool<? extends EventLoopGroup> even
    * GrpcSslContexts}, but options could have been overridden.
    */
   public NettyChannelBuilder sslContext(SslContext sslContext) {
+    checkState(!freezeProtocolNegotiatorFactory,
+               "Cannot change security when using ChannelCredentials");
     if (sslContext != null) {
       checkArgument(sslContext.isClient(),
           "Server SSL context can not be used for client channel");
       GrpcSslContexts.ensureAlpnAndH2Enabled(sslContext.applicationProtocolNegotiator());
     }
-    this.sslContext = sslContext;
+    if (!(protocolNegotiatorFactory instanceof DefaultProtocolNegotiator)) {
+      // Do nothing for compatibility
+      return this;
+    }
+    ((DefaultProtocolNegotiator) protocolNegotiatorFactory).sslContext = sslContext;
```

Comment:
Do you also want to set `((DefaultProtocolNegotiator) protocolNegotiatorFactory).negotiationType` here to TLS?

Reply:
That would change the behavior. I carefully preserved the existing behavior here. These APIs will be removed eventually. It could maybe be updated, but I don't want it updated in this PR, as it increases risk.

Reply:
ok